gh-145559: Add PyUnstable_DumpTraceback() and PyUnstable_DumpTracebackThreads() (#145560)
python/cpython: Doc/c-api/exceptions.rst

   This function returns ``0`` on success, and returns ``-1`` with an

>    exception set on failure.
> 
> .. c:function:: void PyUnstable_DumpTraceback(int fd, PyThreadState *tstate)
> 
>    Write a trace of the Python stack in *tstate* into the file *fd*.
> 
>    This function is safe to use from signal handlers.
> 

Reviewer: ```suggestion

   This function is safe to use from signal handlers.

   The caller does not need to hold an :term:`attached thread state`, nor does
   *tstate* need to be attached.
   
   .. versionadded:: next
```

Reviewer 2: You should add `.. versionadded:: next`.

.. c:function:: const char* PyUnstable_DumpTracebackThreads(int fd, PyInterpreterState *interp, PyThreadState *current_tstate)

   Write the traces of all Python threads in *interp* into the file *fd*.  If

>    *current_state* is not NULL then it will be used to identify what the current

Reviewer: ```suggestion
   *current_tstate* is not ``NULL`` then it will be used to identify what the current
```

>    thread is in the written output. If it is NULL then this function will

Reviewer: ```suggestion
   thread is in the written output. If it is ``NULL``, then this function will
```

   identify the current thread using thread-specific storage.

>    This function will return NULL on success, or an error message on error.

Reviewer: ```suggestion
   This function will return ``NULL`` on success, or an error message on error.
```

>    This function is safe to use from signal handlers.

Reviewer: ```suggestion
   This function is safe to use from signal handlers.
   
   The caller does not need to hold an :term:`attached thread state`, nor does
   *current_tstate* need to be attached.
   
   .. versionadded:: next
```

Reviewer 2: You should add `.. versionadded:: next`.

   This function returns ``0`` on success, and returns ``-1`` with an
   exception set on failure.

.. c:function:: void PyUnstable_DumpTraceback(int fd, PyThreadState *tstate)

   Write a trace of the Python stack in *tstate* into the file *fd*.  The format
   looks like::

      Traceback (most recent call first):
            File "xxx", line xxx in <xxx>
            File "xxx", line xxx in <xxx>
            ...
            File "xxx", line xxx in <xxx>

   This function is meant to debug situations such as segfaults, fatal errors,

>    .etc. The file and function names it outputs are encoded to ASCII with

Reviewer: We try to avoid "etc" because it's bad for translations:
```suggestion
   This function is meant to debug situations such as segfaults, fatal errors,
   and similar. The file and function names it outputs are encoded to ASCII with
```

   backslashreplace and truncated to 500 characters. It writes only the first

>    100 frames, further frames are truncated with the line " ...".

Reviewer: ```suggestion
   100 frames; further frames are truncated with the line ``...``.
```

   This function is safe to use from signal handlers.

   The caller does not need to hold an :term:`attached thread state`, nor does
   *tstate* need to be attached.

   .. versionadded:: next

.. c:function:: const char* PyUnstable_DumpTracebackThreads(int fd, PyInterpreterState *interp, PyThreadState *current_tstate)

   Write the traces of all Python threads in *interp* into the file *fd*.

   If *interp* is ``NULL`` then this function will try to identify the current
   interpreter using thread-specific storage. If it cannot, it will return an
   error.

   If *current_tstate* is not ``NULL`` then it will be used to identify what the
   current thread is in the written output. If it is ``NULL`` then this function
   will identify the current thread using thread-specific storage. It is not an
   error if the function is unable to get the current Python thread state.

   This function will return ``NULL`` on success, or an error message on error.

   This function is meant to debug debug situations such as segfaults, fatal
   errors, .etc. It calls :c:func:`PyUnsafe_DumpTraceback` for each thread. It
   only writes the tracebacks of the first 100 threads, further output is

>    truncated with the line " ...".

Reviewer: ```suggestion
   errors, and similar. It calls :c:func:`PyUnstable_DumpTraceback` for each thread. It
   only writes the tracebacks of the first 100 threads, further output is
   truncated with the line ``...``.
```

   This function returns ``0`` on success, and returns ``-1`` with an
   exception set on failure.

.. c:function:: const char* PyUnstable_DumpTraceback(int fd, PyThreadState *tstate)

   Write a trace of the Python stack in *tstate* into the file *fd*.  The format
   looks like::

      Traceback (most recent call first):
            File "xxx", line xxx in <xxx>
            File "xxx", line xxx in <xxx>
            ...
            File "xxx", line xxx in <xxx>

   This function is meant to debug situations such as segfaults, fatal errors,
   and similar. The file and function names it outputs are encoded to ASCII with
   backslashreplace and truncated to 500 characters. It writes only the first
   100 frames; further frames are truncated with the line ``...``.

   This function will return ``NULL`` on success, or an error message on error.
   It will also write this error message to *fd*.

>    This function is safe to use from signal handlers.

Reviewer 3: These functions aren't safe in general, let alone safe within signal handlers.

* They walk the thread list without any synchronization. 
* They walk interpreter frames which can be concurrently executing, in an inconsistent state, or being actively deleted
* They use thread-local storage, which may not be async-signal-safe, especially if Python is compiled as a shared library.

They make a reasonable effort to print the stack trace, but might themselves crash. You shouldn't use them if you want robust crash handling and reporting.

Reviewer 3: To expand on my earlier comment: I think it's good to make this public as an unstable API, but I think we should set reasonable expectations for the robustness of this in the API docs.

Author: Updated!

   This function returns ``0`` on success, and returns ``-1`` with an
   exception set on failure.

.. c:function:: const char* PyUnstable_DumpTraceback(int fd, PyThreadState *tstate)

   Write a trace of the Python stack in *tstate* into the file *fd*.  The format
   looks like::

      Traceback (most recent call first):
            File "xxx", line xxx in <xxx>
            File "xxx", line xxx in <xxx>
            ...
            File "xxx", line xxx in <xxx>

   This function is meant to debug situations such as segfaults, fatal errors,
   and similar. The file and function names it outputs are encoded to ASCII with
   backslashreplace and truncated to 500 characters. It writes only the first
   100 frames; further frames are truncated with the line ``...``.

   This function will return ``NULL`` on success, or an error message on error.

>    It will also write this error message to *fd*.

Reviewer 2: Hum, I suggest to remove "It will also write this error message to fd." sentence. I'm not sure that it is always a good idea to write the error message to fd.

Author: I thought that was what the [C API discussion](https://github.com/capi-workgroup/decisions/issues/101#issuecomment-4046975462) was about, writing the string result over to the file (when it's non-NULL). Did I misunderstand?

Reviewer 4: The function is supposed to dump debug information to *fd*. It can safely dump this message there, to simplify debugging.

Reviewer 2: PyUnstable_DumpTraceback() should either write the error message to *fd* and return `NULL`, or not write into *fd* and return an error message. Doing both sounds redundant to me.

Currently, `_Py_DumpTracebackThreads()` returns an error message when it knows that it will not be able to output anything useful since it lacks an important data to do its work. For example, it returns an error message if it fails to get the current thread state.

Author: > or not write into fd and return an error message.

I think this is simpler, updated the doc for `PyUnstable_DumpTraceback()`.

Do we still want to do the extra debug logging for `PyUnstable_DumpTracebackThreads()` then, or shall we update that to also only return the `const char*`?
